### ISSUE TITLE
Unnecessary cell-var-from-loop

### ISSUE THREAD

**Issue by sam-s** [pylint-dev/pylint#9677]
### Bug description

When the function closed over a loop variable is only used within the loop, the warning [`cell-var-from-loop`](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/cell-var-from-loop.html) is unnecessary.

### Code

```python
"https://github.com/pylint-dev/pylint/issues/9677"

for x in [1,2,3]:
    def f(a):
        "long and boring"
        print(x)
        print(x+a)
        return x*a
    print(f(2) + f(3) * f(5))
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
a.py:6:14: W0640: Cell variable x defined in loop (cell-var-from-loop)
a.py:7:14: W0640: Cell variable x defined in loop (cell-var-from-loop)
a.py:8:15: W0640: Cell variable x defined in loop (cell-var-from-loop)
```

### Expected behavior

No warning because the closure is only used in the loop

### Pylint version

```shell
pylint 3.2.0
astroid 3.2.0
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
```

### OS / Environment

ubuntu

**Comment by Pierre-Sassoulas** [comment]
Thank you for the triaging @scorphus !
